File: rust/src/security/resource_limits.rs
// #exonware/xwsystem/rust/src/security/resource_limits.rs
//! Generic resource limits and DoS protection.
//! 
//! This module provides resource limit enforcement to prevent DoS attacks
//! and resource exhaustion in any application.


use std::collections::HashMap;

use crate::config::logging_setup::{get_logger};

// Global resource limits registry

// Global resource limits registry

/// Exception raised when resource limits are exceeded.
#[derive(Debug)]
pub struct GenericLimitError {
    message: String,
    source: Option<Box<dyn std::error::Error + Send + Sync + 'static>>,
}

impl std::fmt::Display for GenericLimitError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for GenericLimitError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.source.as_ref().map(|e| e.as_ref() as &(dyn std::error::Error + 'static))
    }
}

impl GenericLimitError {
    pub fn new(message: impl Into<String>) -> Self {
        GenericLimitError {
            message: message.into(),
            source: None,
        }
    }

}

/// Enforces resource limits to prevent DoS attacks.
pub struct ResourceLimits {
    pub component_name: String,
    pub max_depth: i64,
    pub max_resources: i64,
    pub max_path_length: i64,
    _resource_count: i64,
    _depth_count: i64,
}

impl ResourceLimits {
    /// Constructor
    pub fn new(
        component_name: Option<String>,
        max_depth: Option<i64>,
        max_resources: Option<i64>,
        max_path_length: Option<i64>
    ) -> Self {
        Self {
            component_name: component_name.unwrap_or_else(|| "generic".to_string()),
            max_depth: max_depth.unwrap_or(1000),
            max_resources: max_resources.unwrap_or(1000000),
            max_path_length: max_path_length.unwrap_or(1000),
            _resource_count: 0,
            _depth_count: 0,
        }
    }

    /// Check if depth exceeds maximum allowed.
    pub fn check_depth(&mut self, depth: i64) -> Result<(), GenericLimitError> {
        if depth > self.max_depth {
            return Err(GenericLimitError::new(
                format!("Maximum depth {} exceeded: {}", self.max_depth, depth)
            ));
        }
        self._depth_count = self._depth_count.max(depth);
        Ok(())
    }

    /// Check if resource count exceeds maximum allowed.
    pub fn check_resource_count(&mut self, count: i64) -> Result<(), GenericLimitError> {
        if count > self.max_resources {
            return Err(GenericLimitError::new(
                format!("Maximum resource count {} exceeded: {}", self.max_resources, count)
            ));
        }
        self._resource_count = self._resource_count.max(count);
        Ok(())
    }

    /// Check if path length exceeds maximum allowed.
    pub fn check_path_length(&self, path: &str) -> Result<(), GenericLimitError> {
        if path.len() as i64 > self.max_path_length {
            return Err(GenericLimitError::new(
                format!("Path length {} exceeds maximum {}", path.len(), self.max_path_length)
            ));
        }
        Ok(())
    }

    /// Increment resource count and check limit.
    pub fn increment_resource_count(&mut self) -> Result<(), GenericLimitError> {
        self._resource_count += 1;
        self.check_resource_count(self._resource_count)
    }

    /// Get current resource usage statistics.
    pub fn get_stats(&self) -> HashMap<String, serde_json::Value> {
        let mut stats = HashMap::new();
        stats.insert("component".to_string(), serde_json::Value::String(self.component_name.clone()));
        stats.insert("resource_count".to_string(), serde_json::Value::Number(self._resource_count.into()));
        stats.insert("max_depth".to_string(), serde_json::Value::Number(self._depth_count.into()));
        
        let mut limits = HashMap::new();
        limits.insert("max_resources".to_string(), serde_json::Value::Number(self.max_resources.into()));
        limits.insert("max_depth".to_string(), serde_json::Value::Number(self.max_depth.into()));
        limits.insert("max_path_length".to_string(), serde_json::Value::Number(self.max_path_length.into()));
        stats.insert("limits".to_string(), serde_json::to_value(limits).unwrap());
        
        stats
    }
}

// Global resource limits registry
use std::sync::{Arc, Mutex, OnceLock};

static LIMITS_REGISTRY: OnceLock<Mutex<HashMap<String, Arc<Mutex<ResourceLimits>>>>> = OnceLock::new();

fn get_registry() -> &'static Mutex<HashMap<String, Arc<Mutex<ResourceLimits>>>> {
    LIMITS_REGISTRY.get_or_init(|| Mutex::new(HashMap::new()))
}

/// Get resource limits instance for a component.
pub fn get_resource_limits(
    component_name: Option<&str>,
    max_depth: Option<i64>,
    max_resources: Option<i64>,
    max_path_length: Option<i64>
) -> Arc<Mutex<ResourceLimits>> {
    let name = component_name.unwrap_or("generic");
    let mut registry = get_registry().lock().unwrap();
    
    registry.entry(name.to_string())
        .or_insert_with(|| {
            Arc::new(Mutex::new(ResourceLimits::new(
                Some(name.to_string()),
                max_depth,
                max_resources,
                max_path_length
            )))
        })
        .clone()
}

/// Reset resource limits for a component or all components.
pub fn reset_resource_limits(component_name: Option<&str>) {
    let mut registry = get_registry().lock().unwrap();
    if let Some(name) = component_name {
        registry.remove(name);
    } else {
        registry.clear();
    }
}

File: src/exonware/xwsystem/security/resource_limits.py
#exonware/xwsystem/src/exonware/xwsystem/security/resource_limits.py
"""
Generic resource limits and DoS protection.
This module provides resource limit enforcement to prevent DoS attacks
and resource exhaustion in any application.
"""

from pathlib import Path
from typing import Any
from ..config.logging_setup import get_logger


class GenericLimitError(Exception):
    """Exception raised when resource limits are exceeded."""
    pass


class ResourceLimits:
    """Enforces resource limits to prevent DoS attacks."""

    def __init__(
        self,
        component_name: str = "generic",
        max_depth: int = 1000,
        max_resources: int = 1000000,
        max_path_length: int = 1000,
        max_file_size_mb: float = 100.0,
    ):
        self.component_name = component_name
        self.max_depth = max_depth
        self.max_resources = max_resources
        self.max_path_length = max_path_length
        self.max_file_size_mb = max_file_size_mb
        self._logger = get_logger(f"{component_name}.limits")
        self._resource_count = 0
        self._depth_count = 0

    def check_depth(self, depth: int) -> None:
        """
        Check if depth exceeds maximum allowed.
        Args:
            depth: Current depth to check
        Raises:
            GenericLimitError: If depth exceeds limit
        """
        if depth > self.max_depth:
            raise GenericLimitError(f"Maximum depth {self.max_depth} exceeded: {depth}")
        self._depth_count = max(self._depth_count, depth)

    def check_resource_count(self, count: int) -> None:
        """
        Check if resource count exceeds maximum allowed.
        Args:
            count: Current resource count to check
        Raises:
            GenericLimitError: If count exceeds limit
        """
        if count > self.max_resources:
            raise GenericLimitError(
                f"Maximum resource count {self.max_resources} exceeded: {count}"
            )
        self._resource_count = max(self._resource_count, count)

    def check_path_length(self, path: str) -> None:
        """
        Check if path length exceeds maximum allowed.
        Args:
            path: Path to check
        Raises:
            GenericLimitError: If path exceeds limit
        """
        if len(path) > self.max_path_length:
            raise GenericLimitError(
                f"Path length {len(path)} exceeds maximum {self.max_path_length}"
            )

    def check_file_size(self, path: str | Path, max_size_mb: float | None = None) -> bool:
        """
        Validate that a file is within the configured size budget.
        Returns True when the file is within limits or does not exist.
        """
        target = Path(path)
        if not target.exists():
            # Callers frequently validate expected paths before file creation.
            return True
        limit_mb = self.max_file_size_mb if max_size_mb is None else max_size_mb
        size_mb = target.stat().st_size / (1024 * 1024)
        if size_mb > limit_mb:
            raise GenericLimitError(
                f"File size {size_mb:.2f}MB exceeds maximum {limit_mb:.2f}MB"
            )
        return True

    def increment_resource_count(self) -> None:
        """Increment resource count and check limit."""
        self._resource_count += 1
        self.check_resource_count(self._resource_count)

    def get_stats(self) -> dict:
        """Get current resource usage statistics."""
        return {
            "component": self.component_name,
            "resource_count": self._resource_count,
            "max_depth": self._depth_count,
            "limits": {
                "max_resources": self.max_resources,
                "max_depth": self.max_depth,
                "max_path_length": self.max_path_length,
            },
        }
# Global resource limits registry
_limits_registry = {}


def get_resource_limits(
    component_name: str = "generic",
    max_depth: int = 1000,
    max_resources: int = 1000000,
    max_path_length: int = 1000,
) -> ResourceLimits:
    """Get resource limits instance for a component."""
    global _limits_registry
    if component_name not in _limits_registry:
        _limits_registry[component_name] = ResourceLimits(
            component_name, max_depth, max_resources, max_path_length
        )
    return _limits_registry[component_name]


def reset_resource_limits(component_name: str = None) -> None:
    """Reset resource limits for a component or all components."""
    global _limits_registry
    if component_name:
        if component_name in _limits_registry:
            del _limits_registry[component_name]
    else:
        _limits_registry.clear()
